README.md
# sliver_fab

A wrapper widget allowing to place a widget on the edge of FlexibleSpacebar.

![gif](https://raw.githubusercontent.com/MarcinusX/sliver_fab/master/example/screenshots/example.gif)

![gif](https://raw.githubusercontent.com/Matheus-Perez/sliver_fab/master/example/screenshots/exampleTwo.gif)


## [Example usage](https://github.com/MarcinusX/sliver_fab/tree/master/example)
```
class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: new Builder(
        builder: (context) =>
        new SliverFab(
          floatingWidget: FloatingActionButton(
            onPressed: () =>
                Scaffold.of(context).showSnackBar(
                    SnackBar(content: Text("You clicked FAB!"))),
            child: Icon(Icons.add),
          ),
          floatingPosition: FloatingPosition(right: 16),
          expandedHeight: 256.0,
          slivers: <Widget>[
            new SliverAppBar(
              expandedHeight: 256.0,
              pinned: true,
              flexibleSpace: new FlexibleSpaceBar(
                title: new Text("SliverFab Example"),
                background: new Image.asset(
                  "img.jpg",
                  fit: BoxFit.cover,
                ),
              ),
            ),
            new SliverList(
              delegate: new SliverChildListDelegate(
                new List.generate(
                  30,
                      (int index) =>
                  new ListTile(title: new Text("Item $index")),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

## Installing

### 1. Depend on it
Add this to your package's `pubspec.yaml` file:

```
dependencies:
  sliver_fab: "^2.0.0"

```


### 2. Install it
You can install packages from the command line:

with Flutter:

```
$ flutter packages get
```

Alternatively, your editor might support `packages get`. Check the docs for your editor to learn more.

### 3. Import it
Now in your Dart code, you can use:

```
import 'package:sliver_fab/sliver_fab.dart';
```
        

## Contributing

This package was made mostly for use case I needed. If you fill it is missing any features or you have found a bug, **pull requests are very welcome!**


CHANGELOG.md
## [2.0.0]
* Updates the lib for null safety (Chrystoffer Kugler Horochovec)

## [1.0.0]
* **Breaking change:** Parameter `floatingActionButton` changed to `floatingWidget`
* **Breaking change:** Parameter `marginRight` changed to `floatingPosition`
* Fixed scrolling and scaling behavior on iOS.

## [0.1.1]
* Updated environment sdk dependency to Dart 2

## [0.1.0] - Initial release

* Initial release


lib/sliver_fab.dart
library sliver_fab;

import 'package:flutter/material.dart';

/// A helper class if you want a FloatingActionButton to be pinned in the FlexibleAppBar
class SliverFab extends StatefulWidget {
  ///List of slivers placed in CustomScrollView
  final List<Widget> slivers;

  ///FloatingActionButton placed on the edge of FlexibleAppBar and rest of view
  final Widget floatingWidget;

  ///Expanded height of FlexibleAppBar
  final double expandedHeight;

  ///Number of pixels from top from which the [floatingWidget] should start shrinking.
  ///E.g. If your SliverAppBar is pinned, I would recommend this leaving as default 96.0
  ///     If you want [floatingActionButton] to shrink earlier, increase the value.
  final double topScalingEdge;

  ///Position of the widget.
  final FloatingPosition floatingPosition;

  SliverFab({
    required this.slivers,
    required this.floatingWidget,
    this.floatingPosition = const FloatingPosition(right: 16.0),
    this.expandedHeight = 256.0,
    this.topScalingEdge = 96.0,
  });

  @override
  State<StatefulWidget> createState() {
    return new SliverFabState();
  }
}

class SliverFabState extends State<SliverFab> {
  ScrollController? scrollController;

  @override
  void initState() {
    super.initState();
    scrollController = new ScrollController();
    scrollController!.addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    scrollController!.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: <Widget>[
        CustomScrollView(
          controller: scrollController,
          slivers: widget.slivers,
        ),
        _buildFab(),
      ],
    );
  }

  Widget _buildFab() {
    final double defaultFabSize = 56.0;
    final double paddingTop = MediaQuery.of(context).padding.top;
    final double defaultTopMargin = widget.expandedHeight +
        paddingTop +
        (widget.floatingPosition.top ?? 0) -
        defaultFabSize / 2;

    final double scale0edge = widget.expandedHeight - kToolbarHeight;
    final double scale1edge = defaultTopMargin - widget.topScalingEdge;

    double top = defaultTopMargin;
    double scale = 1.0;
    if (scrollController!.hasClients) {
      double offset = scrollController!.offset;
      top -= offset > 0 ? offset : 0;
      if (offset < scale1edge) {
        scale = 1.0;
      } else if (offset > scale0edge) {
        scale = 0.0;
      } else {
        scale = (scale0edge - offset) / (scale0edge - scale1edge);
      }
    }

    return Positioned(
      top: top,
      right: widget.floatingPosition.right,
      left: widget.floatingPosition.left,
      child: new Transform(
        transform: new Matrix4.identity()..scale(scale, scale),
        alignment: Alignment.center,
        child: widget.floatingWidget,
      ),
    );
  }
}

///A class representing position of the widget.
///At least one value should be not null
class FloatingPosition {
  ///Can be negative. Represents how much should you change the default position.
  ///E.g. if your widget is bigger than normal [FloatingActionButton] by 20 pixels,
  ///you can set it to -10 to make it stick to the edge
  final double? top;

  ///Margin from the right. Should be positive.
  ///The widget will stretch if both [right] and [left] are not nulls.
  final double? right;

  ///Margin from the left. Should be positive.
  ///The widget will stretch if both [right] and [left] are not nulls.
  final double? left;

  const FloatingPosition({this.top, this.right, this.left});
}


lib/app_bar_fab.dart
library app_bar_fab;

import 'package:flutter/material.dart';
import 'package:meta/meta.dart';

/// A helper class if you want a FloatingActionButton to be pinned in the FlexibleAppBar
class AppBarFab extends StatefulWidget {
  ///List of slivers placed in CustomScrollView
  final List<Widget> slivers;

  ///FloatingActionButton placed on the edge of FlexibleAppBar and rest of view
  final Widget floatingActionButton;

  ///Expanded height of FlexibleAppBar
  final double expandedHeight;

  ///Right margin of [floatingActionButton]
  final double marginRight;

  ///Number of pixels from top from which the [floatingActionButton] should start shrinking.
  ///E.g. If your SliverAppBar is pinned, I would recommend this leaving as default 96.0
  ///     If you want [floatingActionButton] to shrink earlier, increase the value.
  final double topScalingEdge;

  AppBarFab(
      {@required this.slivers,
      @required this.floatingActionButton,
      this.expandedHeight = 256.0,
      this.marginRight = 16.0,
      this.topScalingEdge = 96.0});

  @override
  State<StatefulWidget> createState() {
    return new AppBarFabState();
  }
}

class AppBarFabState extends State<AppBarFab> {
  ScrollController scrollController;

  @override
  void initState() {
    super.initState();
    scrollController = new ScrollController();
    scrollController.addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    scrollController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return new Stack(
      children: <Widget>[
        new CustomScrollView(
          controller: scrollController,
          slivers: widget.slivers,
        ),
        _buildFab(),
      ],
    );
  }

  Widget _buildFab() {
    final double defaultTopMargin = widget.expandedHeight - 4.0;

    double top = defaultTopMargin;
    double scale = 1.0;
    if (scrollController.hasClients) {
      double offset = scrollController.offset;
      top -= offset;
      if (offset < defaultTopMargin - widget.topScalingEdge) {
        scale = 1.0;
      } else if (offset < defaultTopMargin - widget.topScalingEdge / 2) {
        scale = (defaultTopMargin - widget.topScalingEdge / 2 - offset) /
            (widget.topScalingEdge / 2);
      } else {
        scale = 0.0;
      }
    }

    return new Positioned(
      top: top,
      right: widget.marginRight,
      child: new Transform(
        transform: new Matrix4.identity()..scale(scale, scale),
        alignment: Alignment.center,
        child: widget.floatingActionButton,
      ),
    );
  }
}


example/lib/main.dart
import './exampleOneScreen.dart';
import './exampleTwoScreen.dart';
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Sliver Fab Example',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: AppBar(
        title: Text("Examples"),
      ),
      body: Center(
        child: Column(
          children: <Widget>[
            Padding(
              padding: EdgeInsets.all(16),
              child: RaisedButton(
                child: Text("Example 1"),
                onPressed: () {
                  Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => ExampleOneScreen(),
                      ));
                },
              ),
            ),
            RaisedButton(
              child: Text("Example 2"),
              onPressed: () {
                Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (context) => ExampleTwoScreen(),
                    ));
              },
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/exampleTwoScreen.dart
import 'package:flutter/material.dart';
import 'package:sliver_fab/sliver_fab.dart';

class ExampleTwoScreen extends StatefulWidget {
  @override
  _ExampleTwoScreenState createState() => _ExampleTwoScreenState();
}

class _ExampleTwoScreenState extends State<ExampleTwoScreen> {
  @override
  Widget build(BuildContext context) {
    double media = ((MediaQuery.of(context).size.width - 100) / 2);

    return Scaffold(
      body: new Builder(
        builder: (context) {
          return new SliverFab(
            floatingWidget: Container(
              height: 100,
              width: 100,
              margin: EdgeInsets.only(left: 15.0),
              child: ClipOval(
                child: Image.asset(
                  "img.jpg",
                  fit: BoxFit.fill,
                ),
              ),
              decoration: BoxDecoration(
                color: Color.fromRGBO(255, 255, 255, 0.7),
                shape: BoxShape.circle,
                border: Border.all(
                  color: Color.fromRGBO(255, 255, 255, 0.1),
                  width: 8.0,
                ),
              ),
            ),
            floatingPosition: FloatingPosition(left: media - 10, top: -22),
            expandedHeight: 256.0,
            slivers: <Widget>[
              SliverAppBar(
                expandedHeight: 256.0,
                pinned: true,
                title: Text("Example 2"),
                flexibleSpace: FlexibleSpaceBar(
                  background: Image.asset(
                    "img.jpg",
                    fit: BoxFit.cover,
                  ),
                ),
              ),
              SliverList(
                delegate: SliverChildListDelegate(
                  List.generate(
                    30,
                    (int index) => ListTile(title: Text("Item $index")),
                  ),
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}


example/lib/exampleOneScreen.dart
import 'package:flutter/material.dart';
import 'package:sliver_fab/sliver_fab.dart';

class ExampleOneScreen extends StatefulWidget {
  @override
  _ExampleOneScreenState createState() => _ExampleOneScreenState();
}

class _ExampleOneScreenState extends State<ExampleOneScreen> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: new Builder(
        builder: (context) {
          return new SliverFab(
            floatingWidget: FloatingActionButton(
              onPressed: () => Scaffold.of(context)
                  .showSnackBar(SnackBar(content: Text("You clicked FAB!"))),
              child: Icon(Icons.add),
            ),
            expandedHeight: 256.0,
            slivers: <Widget>[
              new SliverAppBar(
                expandedHeight: 256.0,
                pinned: true,
                flexibleSpace: FlexibleSpaceBar(
                  title: Text("SliverFab Example"),
                  background: Image.asset(
                    "img.jpg",
                    fit: BoxFit.cover,
                  ),
                ),
              ),
              SliverList(
                delegate: SliverChildListDelegate(
                  List.generate(
                    30,
                    (int index) => ListTile(title: Text("Item $index")),
                  ),
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}
